### PR TITLE
JDK-8270308: Arena::Amalloc may return misaligned address on 32-bit

### PR DESCRIPTION
**Pull request by tstuefe**
Hi,

may I please have reviews for this fix. This fixes an issue with arena allocation alignment which can only happen on 32-bit. 

The underlying problem is that even though Arenas offer ways to allocate with different alignment (Amalloc and AmallocWords), allocation alignment is not guaranteed. This sequence will not work on 32-bit:

```
Arena ar;
p1 = ar.AmallocWords();  // return 32bit aligned address
p2 = ar.Amalloc();       // supposed to return 64bit aligned address but does not.
```

This patch is the bare minimum needed to fix the specific problem; I proposed a larger patch before which redid arena alignment handling but it was found too complex: https://github.com/openjdk/jdk/pull/4784 . 

This fix is limited to `Amalloc()` and aligns `_hwm` to be 64bit aligned before allocation. But since chunk boundaries are not guaranteed to be 64-bit aligned either, additional care must be taken to not overflow `_max`.  Since this adds instructions into a hot allocation path, I restricted this code to 32 bit - it is only needed there.

Remaining issues:

- `Amalloc...` align the allocation size in an attempt to ensure allocation alignment. This is not needed nor sufficient, we could just remove that code. I left it untouched to keep the patch minimal. I also left the `// ... for 32 bits this should align _hwm as well.` comment in `Amalloc()` though I think it is wrong.
- The chunk dimensions are not guaranteed to be 64-bit aligned:
	1) Chunk bottom depends on Chunk start. We currently align the header size, but if the chunk starts at an unaligned address, this is not sufficient. It's not a real issue though as long as Chunks are C-heap allocated since malloc alignment is at least 64bit on our 32bit platforms. More of a beauty spot, since this is an implicit assumption which we don't really check.
	2) Chunk top and hence Arena::_max are not guaranteed to be 64-bit aligned either. They depend on the input chunk length, which is not even aligned for the standard chunk sizes used in ChunkPool. And users can hand in any size they want. Fixing this would require more widespread changes to the ChunkPool logic though, so I left it as it is.
	3) similarly, we cannot just align Arena::_max, because that is set in many places and we need to cover all of that; it ties in with Arena rollback as well.

Because of (2) and (3), I had to add the overflow check into `Amalloc()` - any other way to solve this would result in more widespread changes.
	
-----

Tests:

- I tested the provided gtest on both 64-bit and 32-bit platforms (with and without the fix, without it shows the expected problem)
- GHA
- Tests are scheduled at SAP.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270308](https://bugs.openjdk.java.net/browse/JDK-8270308): Arena::Amalloc may return misaligned address on 32-bit

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4835/head:pull/4835` \
`$ git checkout pull/4835`

Update a local copy of the PR: \
`$ git checkout pull/4835` \
`$ git pull https://git.openjdk.java.net/jdk pull/4835/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4835`

View PR using the GUI difftool: \
`$ git pr show -t 4835`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4835.diff">https://git.openjdk.java.net/jdk/pull/4835.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/arena.hpp [hunk 1]
```diff
@@ -133,6 +133,10 @@ class Arena : public CHeapObj<mtNone> {
   // on both 32 and 64 bit platforms. Required for atomic jlong operations on 32 bits.
   void* Amalloc(size_t x, AllocFailType alloc_failmode = AllocFailStrategy::EXIT_OOM) {
     x = ARENA_ALIGN(x);  // note for 32 bits this should align _hwm as well.
+#ifndef LP64 // Since this is a hot path, do this only if really needed.
+    _hwm = ARENA_ALIGN(_hwm);
+    _hwm = MIN2(_hwm, _max); // _max is not guaranteed to be 64 bit aligned.
+#endif // !LP64
```
Comment (coleenp) [inline src/hotspot/share/memory/arena.hpp:139]: This is the change I was thinking of with the conditional for LP64 as well so that 64 bit doesn't pay for the alignment test.

**Review by coleenp — APPROVED** [review]
I think this looks good and minimal.  I'll comment in the other PR.  Maybe I pushed back to hard but these changes are somewhat difficult to review.  It was an impressive amount of work and maybe we are wasting a lot of space in ResourceAreas since many are used to print Symbol->as_C_string().  But even so, they're mostly short lived.
Thank you for doing this change.

**Comment by tstuefe** [comment]
> I think this looks good and minimal. I'll comment in the other PR. Maybe I pushed back to hard but these changes are somewhat difficult to review. It was an impressive amount of work and maybe we are wasting a lot of space in ResourceAreas since many are used to print Symbol->as_C_string(). But even so, they're mostly short lived.
> Thank you for doing this change.

I think it's better this way. This fix is easy to backport. And maybe we do a larger cleanup in later patches. Sorry for running forward with this one - Arenas have been a cause for a lot of frustration in our group, especially before OpenJDK (so, we never could give any changes back).

Realistically, are you guys even interested in cleanups in this area? Since review time is limited, and there are many areas needing attention.

**Comment by tstuefe** [comment]
We crash now in 32-bit with +UseMallocOnly. I'm investigating. This code is fragile :(

**Comment by tstuefe** [comment]
Fixed a bug on 32-bit with +UseMallocOnly. The alignment must never happen in +UseMallocOnly mode since there, the arena is used to keep a packed pointer array and gaps are not helpful. 

Also added a cautionary comment about mixing Amalloc .. calls for arenas with tight packing.

Should we ever clean this up, there are some possible improvements:
- for one, if we really want to keep +UseMallocOnly, the implementation can be simplified by just separating it from the arena and hold a growable pointer array of malloced pointers instead. The problem with keeping these pointers in the area is that when iterating, we need to know the chunk structure. Also, +UseMallocOnly can be used to check if arena handling is broken, but in that case still using the arena does not help.
- We could have "alignment" as a fixed property of the arena and, for arenas where we know they should never allocate with different alignment, verify that in debug.

File: src/hotspot/share/memory/arena.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/memory/arena.hpp:143]: Oh yes, this makes sense that this would have not worked, since UseMallocOnly assumes that the arena is filled with pointers to the malloc memory.

**Review by coleenp — APPROVED** [review]

**Comment by coleenp** [comment]
I don't know if I can speak for everybody but I think cleanups in this area would be welcome if they're incremental and we can plan for time to review them. This hasn't been an area of pain for us (that I know of), but if it has for SAP, maybe cleanups like allowing packing in arenas has a higher priority.
Having a GrowableArray for UseMallocOnly might make sense. I was somewhat surprised by this implementation last week when I first looked at it closely.  Or maybe UseMallocOnly could be removed in favor of memory guards in arena in debug mode, since UseMallocOnly is sparingly tested.  That would be my preference anyway.
Thanks for taking this on and your thoughts on this.

**Comment by tstuefe** [comment]
> I don't know if I can speak for everybody but I think cleanups in this area would be welcome if they're incremental and we can plan for time to review them. This hasn't been an area of pain for us (that I know of), but if it has for SAP, maybe cleanups like allowing packing in arenas has a higher priority.

I'll take a look (with medium priority) whether there are any memory savings possible or whether this is "just" cleanup. If it's the former, more effort may be worth it.

> Having a GrowableArray for UseMallocOnly might make sense. I was somewhat surprised by this implementation last week when I first looked at it closely. Or maybe UseMallocOnly could be removed in favor of memory guards in arena in debug mode, since UseMallocOnly is sparingly tested. That would be my preference anyway.
> Thanks for taking this on and your thoughts on this.

Thanks for the review,

Thomas

File: src/hotspot/share/memory/arena.hpp [hunk 2]
```diff
@@ -134,6 +134,13 @@ class Arena : public CHeapObj<mtNone> {
   void* Amalloc(size_t x, AllocFailType alloc_failmode = AllocFailStrategy::EXIT_OOM) {
     x = ARENA_ALIGN(x);  // note for 32 bits this should align _hwm as well.
     debug_only(if (UseMallocOnly) return malloc(x);)
+    // JDK-8270308: Amalloc guarantees 64-bit alignment and we need to ensure that in case the preceding
+    // allocation was AmallocWords. Note though that padding plays havoc with arenas holding packed arrays,
+    // like HandleAreas. Those areas should never mix Amalloc.. calls with differing alignment.
+#ifndef LP64 // Since this is a hot path, and on 64-bit Amalloc and AmallocWords are identical, restrict this alignment to 32-bit.
+    _hwm = ARENA_ALIGN(_hwm);
+    _hwm = MIN2(_hwm, _max); // _max is not guaranteed to be 64 bit aligned.
```
Comment (kimbarrett) [inline src/hotspot/share/memory/arena.hpp:142]: This may or may not work, depending on what the behavior for `x == 0` is supposed to be.  The current behavior would just return `_hwm` without advancing it.  But if `_max` is not 64bit aligned and `_hwm == _max` then these two steps will end up with `_hwm == _max` again, and a not-64bit aligned result will ensue.  I think better is, for 32bit platforms, allocate an extra word so `_max` can be guaranteed to be 64bit aligned.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/memory/arena.hpp (hunk 2 above)
Comment (kimbarrett) [reply src/hotspot/share/memory/arena.hpp:142]: Or guarantee the chunk size is 64bit aligned and std::max_align_t is also (at least) 64bit aligned (which it almost certainly is), so that the result of malloc will always be (at least) 64bit aligned.

Comment (tstuefe) [reply src/hotspot/share/memory/arena.hpp:142]: I strongly agree that _max and the chunk boundaries should be aligned to 64bit alignment. But as I wrote, that would cause more widespread changes. I would just duplicate the work I did with my first proposal https://github.com/openjdk/jdk/pull/4784. Which already exists, and it does do all you requested. Maybe take a look at that? If you like it, I propose we put this minimal patch through and later use my first PR as a base for a better cleanup.

The current behavior for Amalloc(0) is non-malloc-standard: it returns a non-unique-non-NULL address (since the next allocation returns the same address). When I worked on the first proposal I tried to fix this too. First tried to return NULL, which does not work, callers assume that's an OOM. Then by returning a unique pointer, but I did not like the memory waste. The true fix would be to disallow size==0 and fix callers not to allocate 0 bytes. But I believe that memory allocated with size==0 is not used, since it would be overwritten by the next allocation.

I'll modify the patch and do a size==0->size=1 for 32bit only. Let's hope that does not blow up arena sizes too much.

About std::max_align_t, can we use that? I hard-coded malloc align in a number of places.

Comment (coleenp) [reply src/hotspot/share/memory/arena.hpp]: Can you just assert that size > 0 for Amalloc?

Comment (tstuefe) [reply src/hotspot/share/memory/arena.hpp]: Unfortunately not, since that is actually done. I would have to hunt down callers which do this. That is maybe worthwhile, but would be a larger change.

**Comment by tstuefe** [comment]
@coleenp, @kimbarrett is this current version okay for you?

File: src/hotspot/share/memory/arena.hpp [hunk 3]
```diff
@@ -134,6 +134,15 @@ class Arena : public CHeapObj<mtNone> {
   void* Amalloc(size_t x, AllocFailType alloc_failmode = AllocFailStrategy::EXIT_OOM) {
     x = ARENA_ALIGN(x);  // note for 32 bits this should align _hwm as well.
```
Comment (kimbarrett) [inline src/hotspot/share/memory/arena.hpp:135]: The comment here is stale.

File: src/hotspot/share/memory/arena.hpp (hunk 2 above)
Comment (kimbarrett) [reply src/hotspot/share/memory/arena.hpp]: The latest version doesn't seem any better than the previous.  It could still return a misaligned value in the same circumstance, just more slowly because of the additional branch.

Under what circumstances can _hwm not be 64bit aligned? What needs to be done to prevent it? It looks to me that if Arena::grow were to align its size argument, then it all falls out, assuming std::max_align_t is also appropriately aligned. Add asserts in Chunk allocation. Add a static assert that std::max_align_t is appropriately aligned and leave it to the person porting to the very weird platform to figure out what to do if it fails.  And I think that's sufficient.

(While use of alignof and std::max_align_t are not currently discussed in the HotSpot Style Guide, I would have no objection to their use when the alternative is something more complicated and less robust.  And feel free to propose corresponding post facto changes to style guide.)

**Review by kimbarrett — CHANGES_REQUESTED** [review]
I think this still isn't right.

File: src/hotspot/share/memory/arena.hpp (hunk 3 above)
Comment (tstuefe) [reply src/hotspot/share/memory/arena.hpp:135]: > The comment here is stale.

Yes, I know. See my initial PR text.

I left it in since my first fix was rejected for being too complex, so I wanted to limit this patch to the absolute minimum needed. 

I do not think aligning the allocation size - or insisting on an aligned allocation size as the sister method `AmallocWords` does - is useful at all. Removing the comment just leaves the code, which leaves the reader to ask himself why it's in there. Leaving the comment at least states the reason the original author saw. We should either remove the code and the comment or leave both in place.

**Comment by tstuefe** [comment]
> I think this still isn't right.
>...
> The latest version doesn't seem any better than the previous. It could still return a misaligned value in the same circumstance, just more slowly because of the additional branch.

How? I must be blind.

- If size > 0: Amalloc aligns _hwm to 64-bit and returns a 64-bit aligned address -> ok
- if size == 0: Amalloc does not change the arena, but returns a potentially unaligned pointer - but that is fine too since that pointer cannot be used for storing anything. Certainly nothing needing 64-bit alignment.

What am I overlooking here?

> Under what circumstances can _hwm not be 64bit aligned? What needs to be done to prevent it?

_hwm is misaligned by the caller mixing Amalloc and AmallocWords on 32-bit.

AmallocWords advances _hwm by word size. Then, it is not 64-bit aligned anymore. A subsequent Amalloc() call needs to return a 64-bit aligned pointer, so it needs to align _hwm up. A subsequent AmallocWords() call, however, would be fine with _hwm as it is, so the alignment should really be done at Amalloc().

We cannot just align the allocation size to always be 64-bit, since on an Arena that uses repeated calls to AmallocWords() (e.g., HandleArea) this would create allocation padding and waste memory or even crash if the arena cannot handle padding.

> It looks to me that if Arena::grow were to align its size argument, then it all falls out, assuming std::max_align_t is also appropriately aligned. 

No, since this only ensures the chunk boundaries are properly aligned. The first call to AmallocWords - on a 32-bit platform - will misalign _hwm for potential follow-up calls to Amalloc().

In short, mixing allocation calls with different alignment requirements means the calls with higher alignment guarantees must align. Alternatively, you must ensure that for a given Arena, only one alignment is used - I proposed that alternative in my first patch. Make alignment property of the Arena and only allow allocations with that alignment.

**Comment by kimbarrett** [comment]
Sorry about the typo.  I meant to ask, how can `_max` not be 64bit aligned.  As that is what the comment is claiming can happen, and that is where things can go wrong.  If grow() were to align the allocation size then I don't think that can happen and all would be good.

**Comment by tstuefe** [comment]
> Sorry about the typo. I meant to ask, how can `_max` not be 64bit aligned. As that is what the comment is claiming can happen, and that is where things can go wrong. If grow() were to align the allocation size then I don't think that can happen and all would be good.

Oh, okay. I see what you mean. Arena::_max is not 64-bit aligned currently, but maybe that is easier to fix than I thought:

Arena::_max  depeds on the chunk length (the payload length). I would not want to align Arena::_max itself, because there are too many entry points (all that save restore code directly modifies it). 

But maybe the chunk length itself? The chunk length is either handed in as initsize parameter to `Arena::Arena(MEMFLAGS memflag, size_t init_size);`, which can be anything. But we could align it without problem. We already do, to word size. Could align to 64bit instead.

Or, chunk length is one of the default chunk lengths - `Chunk::size, tinysize` etc. Those are not aligned correctly, since that `slack` offset which we use to keep waste low if libc uses buddy allocation is not aligned correctly. Its 20. If we make it 24, its 64-bit aligned, and all default chunk lengths are 64-bit aligned. 

I think that may work. I have to check. It would certainly be cleaner.

**Comment by tstuefe** [comment]
In this new version, the chunk payload area - and therefore, Arena::_max - is aligned to 64-bit. In order to do this we need to make sure the chunk length given when creating a chunk is 64-bit aligned. There are three places to fix:
- the constants for the cached chunk sizes
- if a caller specifies a custom chunk size as initial size when creating an arena
- when the chunk grows
In all three cases we align now to 64-bit. This works seamlessly.

Tested this new version on x64 and x86 Ubuntu, works. I tested it also with my new gtests (see https://github.com/openjdk/jdk/pull/4898), no problems found.

File: src/hotspot/share/memory/arena.hpp (hunk 3 above)
Comment (kimbarrett) [reply src/hotspot/share/memory/arena.hpp:136]: The length alignment (either by adjustment or by precondition) is what lets the ARENA_ALIGN of _hwm be 32bit only in Amalloc, and not needed in AmallocWords.  Maybe you are suggesting dropping the size alignment and instead unconditionally aligning _hwm as needed for the operation?  Maybe that's the next RFE?

**Review by kimbarrett — CHANGES_REQUESTED** [review]
This looks much better.  But I think you missed a place.

If AmallocWords is called with a non-64bit-aligned value (it could be only 32bit aligned on a 32bit platform), and it calls grow(), grow will call the chunk allocator with that length, which fails the precondition because it's not BytesPerLong aligned. I think grow() needs to call ARENA_ALIGN on the length on 32bit platforms.

That also suggests an additional test.

I like the new comment for Chunk::operator new.

**Comment by tstuefe** [comment]
> This looks much better. But I think you missed a place.
> 
> If AmallocWords is called with a non-64bit-aligned value (it could be only 32bit aligned on a 32bit platform), and it calls grow(), grow will call the chunk allocator with that length, which fails the precondition because it's not BytesPerLong aligned. I think grow() needs to call ARENA_ALIGN on the length on 32bit platforms.

You are right. I even spelled it out in my last comment, but then did not fix it.

I added a test, saw that it fires on 32-bit, then fixed Arena::grow(). I also limited all new tests to 32-bit to save some cycles on 64-bit.

One remaining worry I have is that when mixing Amalloc and AmallocWords now we align correctly, that is fine. But if the Arena cannot handle gaps - e.g. HandleArea - difficult to analyze crashes can happen on 32-bit. The original cause would be that the author mixes Amalloc and AmallocWords for an Arena where he should stick to one alignment only.

My first patch filled alignment gaps in debug with a pattern, to trip off the analyzer. What do you think, should I do this here too?

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by tstuefe** [comment]
Thanks, @kimbarrett !

@coleenp are you ok with the latest version too?

File: src/hotspot/share/memory/arena.hpp [hunk 4]
```diff
@@ -52,11 +52,12 @@ class Chunk: CHeapObj<mtChunk> {
   enum {
     // default sizes; make them slightly smaller than 2**k to guard against
     // buddy-system style malloc implementations
+    // Note: please keep these constants 64-bit aligned.
```
Comment (coleenp) [inline src/hotspot/share/memory/arena.hpp:55]: Can you add a static_assert(is_aligned(slack, ARENA_ALIGNMENT)?

File: src/hotspot/share/memory/arena.hpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/memory/arena.hpp:136]: Oh please no more RFEs unless Kim can review them.  This turned out much more difficult than we thought it would be!
This comment is true though, right?  For 32 bits if you do an AmallocWords then Amalloc, the _hwm will then be aligned to ARENA_ALIGN, ie 64 bits.  That's ok.  That seems like just what we want.  Although it seems like this latter comment says what I think this does better.  I'm fine with leaving the comment though.  Makes you think about it anyway.

File: src/hotspot/share/memory/arena.cpp [hunk 5]
```diff
@@ -171,13 +171,30 @@ class ChunkPoolCleaner : public PeriodicTask {
...
+  // - requested_size = sizeof(Chunk)
+  // - length = payload size
+  // We must ensure that the boundaries of the payload (C and D) are aligned to 64-bit:
+  //
+  // +-----------+--+--------------------------------------------+
+  // |           |g |                                            |
+  // | Chunk     |a |               Payload                      |
+  // |           |p |                                            |
+  // +-----------+--+--------------------------------------------+
+  // A           B  C                                            D
```
Comment (coleenp) [inline src/hotspot/share/memory/arena.cpp:185]: Hooray for ascii art!

File: src/hotspot/share/memory/arena.cpp [hunk 6]
```diff
@@ -171,13 +171,30 @@ class ChunkPoolCleaner : public PeriodicTask {
...
+  //   64-bit aligned on all our platforms, including 32-bit.
+  // - sizeof(Chunk) (B) may not be aligned to 64-bit, and we have to take that into
+  //   account when calculating the Payload bottom (C) (see Chunk::bottom())
+  // - the payload size (length) must be aligned to 64-bit, which takes care of 64-bit
+  //   aligning (D)
+
+  assert(sizeofChunk == sizeof(Chunk), "weird request size");
+  assert(is_aligned(length, BytesPerLong), "chunk payload length not 64-bit aligned: "
+                                           SIZE_FORMAT ".", length);
+  size_t bytes = ARENA_ALIGN(sizeofChunk) + length;
```
Comment (coleenp) [inline src/hotspot/share/memory/arena.cpp:197]: Are Chunk::size, medium_size, init_size and tiny size all 64 bit aligned?  I see that they're aligned here again but it seems like those sizes being invariantly aligned to 64 bit would prevent future bugs (if they're not already).  Also maybe needing a static_assert ?

File: test/hotspot/gtest/memory/test_arena.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (coleenp) [inline test/hotspot/gtest/memory/test_arena.cpp:2]: Did you want an SAP copyright?

File: test/hotspot/gtest/memory/test_arena.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "memory/arena.hpp"
+#include "unittest.hpp"
+
+#ifndef LP64
+// These tests below are about alignment issues when mixing Amalloc and AmallocWords.
+// Since on 64-bit these APIs offer the same alignment, they only matter for 32-bit.
```
Comment (coleenp) [inline test/hotspot/gtest/memory/test_arena.cpp:30]: Thank you so much for testing this on 32 bit platforms.

File: test/hotspot/gtest/memory/test_arena.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+TEST_VM(Arena, mixed_alignment_allocation) {
+  // Test that mixed alignment allocations work and provide allocations with the correct
+  // alignment
+  Arena ar(mtTest);
+  void* p1 = ar.AmallocWords(BytesPerWord);
+  void* p2 = ar.Amalloc(BytesPerLong);
+  ASSERT_TRUE(is_aligned(p1, BytesPerWord));
+  ASSERT_TRUE(is_aligned(p2, BytesPerLong));
```
Comment (coleenp) [inline test/hotspot/gtest/memory/test_arena.cpp:39]: Should BytesPerLong in this test be ARENA_AMALLOC_ALIGNMENT?

File: src/hotspot/share/memory/arena.cpp [hunk 10]
```diff
@@ -188,6 +205,7 @@ void* Chunk::operator new (size_t requested_size, AllocFailType alloc_failmode,
      if (p == NULL && alloc_failmode == AllocFailStrategy::EXIT_OOM) {
        vm_exit_out_of_memory(bytes, OOM_MALLOC_ERROR, "Chunk::new");
      }
+     assert(is_aligned(p, BytesPerLong), "Chunk start address not malloc aligned?");
```
Comment (coleenp) [inline src/hotspot/share/memory/arena.cpp:208]: Should this be ARENA_AMALLOC_ALIGNMENT, in case we have to change it back to 128 bits for some reason?

**Review by coleenp — CHANGES_REQUESTED** [review]
These are small comments but maybe important.

**Comment by coleenp** [comment]
> My first patch filled alignment gaps in debug with a pattern, to trip off the analyzer. What do you think, should I do this here too?
Maybe. Aren't arena's freed with a zap pattern so you'd find bugs that way? Anyway, it might be better as a separate RFE and new round of testing.
Thanks for default aligning the chunks though because this is what one would expect without looking at the code more closely.

File: src/hotspot/share/memory/arena.hpp (hunk 4 above)
Comment (tstuefe) [reply src/hotspot/share/memory/arena.hpp:55]: Sure!

File: src/hotspot/share/memory/arena.cpp (hunk 6 above)
Comment (tstuefe) [reply src/hotspot/share/memory/arena.cpp:197]: > Are Chunk::size, medium_size, init_size and tiny size all 64 bit aligned? I see that they're aligned here again but it seems like those sizes being invariantly aligned to 64 bit would prevent future bugs (if they're not already). Also maybe needing a static_assert ?

They are all aligned now - and you are right, I can add static asserts to assert that. I do not align them here, I just assert that they are aligned (via assert(is_aligned(length)).

The only thing actively I align here is the header size aka sizeof(Chunk). I tried to force the compiler to generate Chunk as 64-bit aligned structure but gave up after fiddling with pragmas for half an hour.

File: test/hotspot/gtest/memory/test_arena.cpp (hunk 7 above)
Comment (tstuefe) [reply test/hotspot/gtest/memory/test_arena.cpp:2]: I could add one. Probably should, thanks for noticing.

File: test/hotspot/gtest/memory/test_arena.cpp (hunk 8 above)
Comment (tstuefe) [reply test/hotspot/gtest/memory/test_arena.cpp:30]: Sure. I run Ubuntu 20.4, with Debian multi-arch it's beautifully easy to have 32-bit and 64-bit compilers side by side.

File: test/hotspot/gtest/memory/test_arena.cpp (hunk 9 above)
Comment (tstuefe) [reply test/hotspot/gtest/memory/test_arena.cpp:39]: Yes, I can do that.

File: src/hotspot/share/memory/arena.cpp (hunk 10 above)
Comment (tstuefe) [reply src/hotspot/share/memory/arena.cpp:208]: Yes, I can do that too. I should try to set it to 128bits and test it. 

With 128 bit we may run into alignment problems with 32-bit platforms, since we rely on ARENA_AMALLOC_ALIGNMENT being <= malloc alignment, and on 32-bit I believe malloc alignment is just 64 bit. Not a showstopper, but may require a bit more care.

**Comment by tstuefe** [comment]
> > My first patch filled alignment gaps in debug with a pattern, to trip off the analyzer. What do you think, should I do this here too?
> Maybe. Aren't arena's freed with a zap pattern so you'd find bugs that way? 

Yes, you are right, maybe that's sufficient. I thought that the zap pattern would be overwritten with the first allocation wave (allocate a bunch, reset-to-mark, allocate again) but I see reset-to-mark - ultimately Chunk::chop() - also zaps, so this may be fine.

> Anyway, it might be better as a separate RFE and new round of testing.

Yes. But the default pattern may be fine already.

> Thanks for default aligning the chunks though because this is what one would expect without looking at the code more closely.

Sure!

**Comment by tstuefe** [comment]
New version:
- added static asserts to test that the default chunk sizes are now properly aligned (since Chunk::slack is used to compute those sizes, this also tests Chunk::slack)
- changed a number of literal checks for 64-bit aligned-ness to ARENA_AMALLOC_ALIGNMENT
- added SAP copyright

File: src/hotspot/share/memory/arena.cpp (hunk 10 above)
Comment (coleenp) [reply src/hotspot/share/memory/arena.cpp]: I was wondering if it would make an interesting test mode.

**Review by coleenp — APPROVED** [review]
Looks great.  Thank you!

**Review by kimbarrett — APPROVED** [review]
Looks even better after dealing with Coleen's recent comments.

**Comment by tstuefe** [comment]
Thanks Coleen & Kim!

/integrate
